scipy/stats/_stats_py.py

```diff
@@ -9440,3 +9442,88 @@ def rankdata(a, method='average', *, axis=None, nan_policy='propagate'):
         result[nan_indexes] = np.nan
 
     return result
+
+
+# TODO: axis parameter
+# TODO: add _axis_nan_policy_factory decorator
+def expectile(a, alpha=0.5, axis=0, dtype=None, weights=None):
+    r"""Compute the expectile along the specified axis.
+
+    The expectile at level alpha of the array :math:`a_i` associated to
+    weights :math:`w_i` is the unique solution :math:`t` of:
+
+    .. math::
+
+        \alpha \sum_{i=1}^n w_i (y_i - t)_+ =
+            (1 - \alpha) \sum_{i=1}^n w_i (t - y_i)_+ \,.
+
+    Expectiles are a generalization of the expectation in the same way as
+    quantiles are a generalization of the median. The expectile at level
+    `alpha = 0.5` is the mean (average).
+
+    Parameters
+    ----------
+    a : array_like
+        Input array or object that can be converted to an array.
+    alpha : float
+        The level of the expectile; `alpha=0.5` give the mean.
+    axis : int or None, optional
+        Axis along which the geometric mean is computed. Default is 0.
+        If None, compute over the whole array `a`.
+    dtype : dtype, optional
+        Type to which the input arrays are cast before the calculation is
+        performed.
+    weights : array_like, optional
+        The `weights` array must be broadcastable to the same shape as `a`.
+        Default is None, which gives each value a weight of 1.0.
+
+    Returns
+    -------
+    expectile : ndarray
+        See `dtype` parameter above.
+
+    See Also
+    --------
+    numpy.mean : Arithmetic average
+    numpy.quantile : Quantile
+
+    References
+    ----------
+    .. [1] W. K. Newey and J. L. Powell (1987), "Asymmetric Least Squares
+           Estimation and Testing," Econometrica, 55, 819-847.
+
+    Examples
+    --------
+    >>> from scipy.stats import gmean
+    >>> gmean([1, 4])
+    2.0
+    >>> gmean([1, 2, 3, 4, 5, 6, 7])
+    3.3800151591412964
+    >>> gmean([1, 4, 7], weights=[3, 1, 3])
+    2.80668351922014
+
+    """
+    if alpha < 0 or alpha > 1:
+        raise ValueError(
+            "The expectile level alpha must be in the range [0, 1]."
+        )
+    a = np.asarray(a, dtype=dtype)
+
+    if weights is not None:
+        weights = np.asarray(weights, dtype=dtype)
+
+    def first_order(t):
+        return (
+            alpha * np.average(np.fmax(0, a - t), weights=weights)
+            - (1 - alpha) * np.average(np.fmax(0, t - a), weights=weights)
+        )
+
+    if alpha >= 0.5:
+        x0 = np.average(a, weights=weights)
+        x1 = np.amax(a)
+    else:
+        x1 = np.average(a, weights=weights)
+        x0 = np.amin(a)
+
+    res = root_scalar(first_order, x0=x0, x1=x1)
```

Comment:
It would be good to note which equation this correspond with.

Reply:
I added a reference. But there is none for including `weights`. This pretty much follows from analogy with the mean and quantiles.

Reply:
The original formula comes from the first-order condition for minimizing a quadratic function. I suppose if weights were added to the quadratic (if it makes sense to do so in the original context?), the minimizer would involve weighted averages like you have here? If that's the case, I think you'd need to add a restriction to the signs of the weights, no? In any case, if there is not a reference, there needs to be an argument for adding them.

I'd also suggest prioritizing `weights` vs `axis`/`nan_policy`. It will be more challenging to add those if this has `weights`.

Reply:
It is difficult to weigh up `weights` vs `axis`, I prefer `weights` a little more.
You are right about the minimizer. One can define expectiles as `argmin_t E[loss(t, Y)]` for certain loss functions.

I would argue in favor of including weights, which becomes handy for aggregated data. One can see them as part of the empirical distribution `F` when taking the expectation in `E[loss(t, Y)]`.

Reply:
What I was worried about is that although the first-order conditions may still be valid, the second-order conditions can change, e.g. if all weights are negative, the minimizer becomes the maximizer. 

Here's the test I had in mind to add with (and without) `weights`:
```python3
import numpy as np
from scipy import stats, optimize
rng = np.random.default_rng(1638083107694713882823079058616272161)

p = 0.33
n = 10
weights = rng.random(n)
x = rng.random(n)

def fun(u):
    w = np.full_like(x, fill_value=p)
    w[x < u] = 1-p
    s = w * weights * (x - u)**2
    return np.sum(s)
    
res = optimize.minimize_scalar(fun, bracket=(np.min(x), np.max(x)))
print(res.x)  # 0.4237264854407342
print(stats.expectile(x, alpha=p, weights=weights))  # 0.4237264854407448
```

But when all weights are negative, that's the same as changing the sign of the objective function, so `expectile` gives the maximum of this function. 

I guess `np.average` does the same thing.
```python3
np.average(x, weights=np.ones_like(x))  # 0.5180018870779544
np.average(x, weights=-np.ones_like(x))  # 0.5180018870779544
```
but still, I think it's worth thinking about for a moment. 
@h-vetinari @steppi what do you think? Weights aren't mentioned in the literature, so there's a question about how they should be defined (if we are going to support them). I'd think that they would be weights applied to the terms of a function that the expectile minimizes, but that's not exactly what happens in the code.

Reply:
Statistically speaking, expectiles (as well as the mean and quantiles) are properties of a distribution function `F`, e.g. `T(F)`. The empirical version arises by pluggin in the empirical distribution `\hat{F}` of a given sample `x`.
In my point of view, sample weights/case weights are defined on this distribution in a relative sense; for integer valued weights, it can be interpreted as repeated observation.
The weighted empirical cdf is `\hat{F}(t) = 1/sum(w) * sum_i w_i 1_{x_i <= t}`.

Implied assumptions and properties:
1. Weights are non-negative.
2. A weight of 2 is equivalent to duplicating this observation.
3. `\hat{F}` as well as `T(\hat{F})` are invariant under scalar rescaling: `w -> c*w` for `c > 0`
4. The expected value (population level) reads as weighted average on a sample level in the defining equations:
  - Minimisation with strict consistent scoring function `S` for `T`: `T(F) = argmin_z E_F(S(z, X))` for all `X~F`
    Sample level: `T(\hat{F}) = argmin_z 1/sum(w) sum_i w_i S(z, x_i)`
  - First order condition / strict identification function `V` for `T`: `T(F)` is solution of `E_F(V(z, X)) = 0` for all `X~F`
    Sample level: `T(\hat{F})` is solution of `1/sum(w) sum_i w_i V(z, x_i) = 0`

Note: For expectiles, the solution is unique (and yes, we need to restict to some class of distribution functions `F`, like finite first moment)

Reply:
BTW GitHub plays with $\LaTeX$ now.

Reply:
That is fantastic news. It would have helped me many times!